library/src/main/java/com/novoda/downloadmanager/lib/DownloadManager.java

```diff
@@ -472,21 +481,38 @@ public int removeDownloads(long... ids) {
      * it was running, and it will no longer be accessible through the download manager.
      * If there are any downloaded files, partial or complete, they will be deleted.
      *
-     * @param ids the IDs of the batches to remove
+     * @param batchesIds the IDs of the batches to remove
      * @return the number of batches actually removed
      */
-    public int removeBatches(long... ids) {
-        if (ids == null || ids.length == 0) {
-            throw new IllegalArgumentException("called with nothing to remove. input param 'ids' can't be null");
+    public int removeBatches(long... batchesIds) {
+        if (batchesIds == null || batchesIds.length == 0) {
+            throw new IllegalArgumentException("called with nothing to remove. input param 'batchesIds' can't be null");
         }
-        ContentValues values = new ContentValues();
-        values.put(DownloadContract.Batches.COLUMN_DELETED, 1);
-        // if only one id is passed in, then include it in the uri itself.
-        // this will eliminate a full database scan in the download service.
-        if (ids.length == 1) {
-            return contentResolver.update(ContentUris.withAppendedId(downloadsUriProvider.getBatchesUri(), ids[0]), values, null, null);
+
+        setDeletingStatusFor(batchesIds);
+        return markBatchesBeDeleted(batchesIds);
+    }
+
+    private void setDeletingStatusFor(long[] batchesIds) {
```

Comment:
One is to add a flag in the database that this batch will be deleted by the DownloadService thread, so it marked now, will be deleted in the future.

This one sets **now** the new status, that will trigger a content observer update and everyone listening for the status changes in the database will be notified immediately. This is the part of the code that updates the status immediately after clicking on it instead of having to wait for 3s for the thread to pick it up and do the update.